rust/derive/src/stringenum.rs

```diff
@@ -0,0 +1,127 @@
+                        if let Ok(value) = li.base10_parse::<T>() {
+                            values.push(value);
+                        } else {
+                            panic!("EnumString requires explicit {}", ustr);
+                        }
+                    } else {
+                        panic!("EnumString requires explicit literal integer");
+                    }
+                } else {
+                    panic!("EnumString requires explicit literal");
+                }
+            } else {
+                if let syn::Fields::Unnamed(f) = &v.fields {
+                    if default_seen || f.unnamed.len() != 1 {
+                        panic!("EnumString requires explicit values or one Unknown({})", ustr);
+                    }
+                    if v.ident.to_string() != "Unknown" {
+                        panic!("EnumString default case must be Unknown({})", ustr);
+                    }
+                    if let syn::Type::Path(p) = &f.unnamed[0].ty {
+                        if p.path.segments.len() != 1 {
+                            panic!("EnumString requires explicit values or one Unknown({})", ustr);
+                        }
+                        if p.path.segments[0].ident.to_string() != ustr {
+                            panic!("EnumString default case must be Unknown({})", ustr);
+                        }
+                    } else {
+                        panic!("EnumString requires explicit values or one Unknown({})", ustr);
+                    }
+                    default_seen = true;
+                } else {
+                    panic!("EnumString requires explicit values or one Unknown({})", ustr);
+                }
+            }
+        }
+    } else {
+        panic!("EnumString can only be derived for enums");
+    }
+
+    let utype_str = syn::Ident::new(&ustr, proc_macro2::Span::call_site());
+
+    let expanded = quote! {
+        impl #name {
+            pub(crate) fn from_u(v: #utype_str) -> Self {
+                match v {
+                    #( #values => #name::#fields ,)*
+                    _ => #name::Unknown(v),
+                }
+            }
+            pub(crate) fn into_u(&self) -> #utype_str {
+                match *self {
+                    #( #name::#fields => #values ,)*
+                    #name::Unknown(v) => v,
+                }
+            }
+            pub(crate) fn to_string(&self) -> String {
+                match *self {
+                    #( #name::#fields => #names.to_string() ,)*
+                    #name::Unknown(v) => format!("unknown-{}", v),
+                }
+            }
+            pub(crate) fn from_str(s: &str) -> Option<#name> {
+                match s {
+                    #( #names => Some(#name::#fields) ,)*
+                    _ => None
+                }
+            }
+            pub(crate) fn to_detect_ctx(s: &str) -> Option<DetectUintData<#utype_str>> {
```

Comment:
I think this derive needs a new name as it does more than just to/from strings.

It is helpful when a derive generates the implementation of a trait, like `AppLayerEvent` as it gives you a place to document the interface that is being generated. `AppLayerFrameType` is another example.

Reply:
More on using a trait.  If we create a trait, `StringEnumU8`, then we should be able to get this method out of the derive, as I think derives should be the minimal amount of code needed.

We could then do something like:
```
impl From<&dyn StringEnumU8> for DetectUintData {
    type Error = ();
    fn try_from(value: &dyn MyTrait) -> Result<Self, Self::Error> {
    }
}
```

allowing anything implementing StringEnumU8 to be converted to a DetectUintData through standard interfaces.

Reply:
Ok to make a trait.

Not getting your second comment
How can we get the method out of the derive ?

Reply:
> How can we get the method out of the derive ?

Just implement the From trait for DetectUintData alongside DetectUintData.

Reply:
I think I get what you mean.

But I do not really do `impl From<&dyn StringEnumU8> for DetectUintData {` because I resort to generic integer parsing if this is not an enum string...

So I think the prototype `fn to_detect_ctx(s: &str) -> Option<DetectUintData<T>>` is right as it takes a string as input